File: android-compose/app/src/main/java/com/motchill/androidcompose/feature/detail/DetailPresentation.kt
package com.motchill.androidcompose.feature.detail

import com.motchill.androidcompose.core.storage.PlaybackProgressSnapshot
import com.motchill.androidcompose.domain.model.MovieDetail

enum class DetailSectionTab {
    episodes,
    synopsis,
    information,
    classification,
    gallery,
    related,
}

val DetailSectionTab.label: String
    get() = when (this) {
        DetailSectionTab.episodes -> "Episodes"
        DetailSectionTab.synopsis -> "Synopsis"
        DetailSectionTab.information -> "Information"
        DetailSectionTab.classification -> "Classification"
        DetailSectionTab.gallery -> "Gallery"
        DetailSectionTab.related -> "Related"
    }

data class DetailUiState(
    val isLoading: Boolean = true,
    val errorMessage: String? = null,
    val detail: MovieDetail? = null,
    val selectedTab: DetailSectionTab? = null,
    val isLiked: Boolean = false,
    val episodeProgressById: Map<Int, PlaybackProgressSnapshot> = emptyMap(),
) {
    val availableTabs: List<DetailSectionTab>
        get() = detail?.availableTabs.orEmpty()

    val effectiveSelectedTab: DetailSectionTab
        get() = selectedTab?.takeIf { it in availableTabs }
            ?: detail?.let(::defaultDetailTab)
            ?: DetailSectionTab.synopsis
}

val MovieDetail.availableTabs: List<DetailSectionTab>
    get() {
        val tabs = mutableListOf<DetailSectionTab>()
        if (episodes.isNotEmpty()) tabs += DetailSectionTab.episodes
        if (description.trim().isNotEmpty()) tabs += DetailSectionTab.synopsis
        if (hasInformation()) tabs += DetailSectionTab.information
        if (countries.isNotEmpty() || categories.isNotEmpty()) {
            tabs += DetailSectionTab.classification
        }
        if (photoUrls.isNotEmpty() || previewPhotoUrls.isNotEmpty()) {
            tabs += DetailSectionTab.gallery
        }
        if (relatedMovies.isNotEmpty()) tabs += DetailSectionTab.related
        return tabs
    }

fun defaultDetailTab(detail: MovieDetail): DetailSectionTab {
    val tabs = detail.availableTabs
    if (tabs.isEmpty()) return DetailSectionTab.synopsis
    if (DetailSectionTab.episodes in tabs) return DetailSectionTab.episodes
    return tabs.first()
}

private fun MovieDetail.hasInformation(): Boolean {
    return director.trim().isNotEmpty() ||
        castString.trim().isNotEmpty() ||
        showTimes.trim().isNotEmpty() ||
        moreInfo.trim().isNotEmpty() ||
        trailer.trim().isNotEmpty() ||
        statusRaw.trim().isNotEmpty() ||
        statusText.trim().isNotEmpty()
}

File: swiftUI/Features/Detail/DetailPresentation.swift
import Foundation

enum DetailSectionTab: Hashable, Sendable {
    case episodes
    case synopsis
    case information
    case classification
    case gallery
    case related

    var label: String {
        switch self {
        case .episodes: return "Episodes"
        case .synopsis: return "Synopsis"
        case .information: return "Information"
        case .classification: return "Classification"
        case .gallery: return "Gallery"
        case .related: return "Related"
        }
    }
}

enum DetailScreenState: Equatable, Sendable {
    case idle
    case loading
    case loaded
    case error(message: String)
}

extension PhucTvMovieDetail {
    var availableTabs: [DetailSectionTab] {
        var tabs: [DetailSectionTab] = []
        if !episodes.isEmpty { tabs.append(.episodes) }
        if !description.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty { tabs.append(.synopsis) }
        if hasInformation { tabs.append(.information) }
        if !countries.isEmpty || !categories.isEmpty { tabs.append(.classification) }
        if !photoUrls.isEmpty || !previewPhotoUrls.isEmpty { tabs.append(.gallery) }
        if !relatedMovies.isEmpty { tabs.append(.related) }
        return tabs
    }

    var hasInformation: Bool {
        !director.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ||
            !castString.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ||
            !showTimes.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ||
            !moreInfo.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ||
            !trailer.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ||
            !statusRaw.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ||
            !statusText.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    }

    var defaultTab: DetailSectionTab {
        let tabs = availableTabs
        if tabs.isEmpty { return .synopsis }
        if tabs.contains(.episodes) { return .episodes }
        return tabs.first ?? .synopsis
    }
}
